registration/include/pcl/registration/impl/correspondence_estimation_backprojection.hpp

```diff
@@ -188,98 +144,49 @@ CorrespondenceEstimationBackProjection<PointSource, PointTarget, NormalT, Scalar
   unsigned int nr_valid_correspondences = 0;
   int target_idx = 0;
 
-  // Check if the template types are the same. If true, avoid a copy.
-  // Both point types MUST be registered using the POINT_CLOUD_REGISTER_POINT_STRUCT
-  // macro!
-  if (isSamePointType<PointSource, PointTarget>()) {
-    PointTarget pt;
-    // Iterate over the input set of source indices
-    for (const auto& idx_i : (*indices_)) {
-      tree_->nearestKSearch((*input_)[idx_i], k_, nn_indices, nn_dists);
-
-      // Among the K nearest neighbours find the one with minimum perpendicular distance
-      // to the normal
-      float min_dist = std::numeric_limits<float>::max();
-
-      // Find the best correspondence
-      for (std::size_t j = 0; j < nn_indices.size(); j++) {
-        float cos_angle = (*source_normals_)[idx_i].normal_x *
-                              (*target_normals_)[nn_indices[j]].normal_x +
-                          (*source_normals_)[idx_i].normal_y *
-                              (*target_normals_)[nn_indices[j]].normal_y +
-                          (*source_normals_)[idx_i].normal_z *
-                              (*target_normals_)[nn_indices[j]].normal_z;
-        float dist = nn_dists[j] * (2.0f - cos_angle * cos_angle);
-
-        if (dist < min_dist) {
-          min_dist = dist;
-          min_index = static_cast<int>(j);
-        }
+  // Iterate over the input set of source indices
+  for (const auto& idx_i : (*indices_)) {
+    // Check if the template types are the same. If true, avoid a copy.
+    // Both point types MUST be registered using the POINT_CLOUD_REGISTER_POINT_STRUCT
+    // macro!
+    const auto& pt = detail::pointCopyOrRef<PointSource, PointTarget, decltype(idx_i)>(
```

Comment:
Do we need the long form here?

Reply:
revisited

Reply:
while it is possible to deduce `Index` type, it's not true for `PointSource` (second type parameter): type cannot be infered as it is used in nested-type-specifier